### PR TITLE
Fix: function definition is marked dllimport

### LINKED ISSUE #209: Compiling error: function definition is marked dllimport
Hi,

When fastcdr is compiled and installed, I continue to build fastrtps, (compile on Windows and set `-DBUILD_SHARED_LIBS=ON`),

then I got lots of similar errors:

```sh
C:/ros/jazzy/install/include/fastcdr/Cdr.h:2792:21: error: function 'eprosima::fastcdr::Cdr& eprosima::fastcdr::Cdr::deserialize_type(eprosima::fastcdr::EncodingAlgorithmFlag, std::function<bool(eprosima::fastcdr::Cdr&, const eprosima::fastcdr::MemberId&)>)' definition is marked dllimport
 2792 |     Cdr_DllAPI Cdr& deserialize_type(
      |                     ^~~~~~~~~~~~~~~~
```
the reason is that when compiled as shared lib (dll on windows), `Cdr_DllAPI` is defined as `__declspec( dllexport )` when compiling fastcdr and `__declspec( dllimport )` when compiling fastrtps and includes fastcdr as a shared lib.

Then the included header file (for fastftps), is like this
```c++
//Cdr.h
    Cdr_DllAPI Cdr& deserialize_type(
            EncodingAlgorithmFlag type_encoding,
            std::function<bool (Cdr&, const MemberId&)> functor)
    {
        return (this->*deserialize_type_)(type_encoding, functor);
    }
```
the error message says that the function is marked as `dllimport` but it has a function body (definition) at the same time, which is an error.

The suggested change is to move the function body/definition from .h file to .cpp file.

I'd like to create a PR if the suggestion above is acceptable.

### PR DESCRIPTION
**Pull request by felixf4xu**
Fix #209

### CODE REVIEW

**Review by JesusPoderoso — APPROVED** [review]
LGTM!
Thanks for your contribution @felixf4xu! 🚀

File: src/cpp/Cdr.cpp [hunk 1]
```diff
@@ -2151,6 +2193,40 @@ Cdr& Cdr::deserialize_array(
     throw NotEnoughMemoryException(NotEnoughMemoryException::NOT_ENOUGH_MEMORY_MESSAGE_DEFAULT);
 }
 
+Cdr& Cdr::begin_serialize_type(
+        Cdr::state& current_state,
+        EncodingAlgorithmFlag type_encoding)
+{
+    return (this->*begin_serialize_type_)(current_state, type_encoding);
+}
+
```
Comment (JesusPoderoso) [inline src/cpp/Cdr.cpp:2202]:
```suggestion
```

File: src/cpp/Cdr.cpp [hunk 2]
```diff
@@ -2151,6 +2193,40 @@ Cdr& Cdr::deserialize_array(
...
+    if (next_member_id_ != MEMBER_ID_INVALID)
+    {
+        throw exception::BadParamException("Member id already set and not encoded");
+    }
+
+    next_member_id_ = member_id;
+    return *this;
+}
+
+
```
Comment (JesusPoderoso) [inline src/cpp/Cdr.cpp:2229]:
```suggestion
```

**Review by JesusPoderoso — CHANGES_REQUESTED** [review]
Just a nitpick to pass our linter

**Review by JesusPoderoso — APPROVED** [review]

**Comment by felixf4xu** [comment]
Thanks for the review.

At the same time, I have a question about the compiling switches. The issue only happens when fastcdr is built as a shared dll (on Windows), in ros2, it's `BUILD_SHARED_LIBS`. I'm just curious, this project has all code ready (the difinition of `Cdr_DllAPI` and all the other stuff) but this issue show that maybe it's never compiled on Windows as shared dll.

ROS2 actually supports Windows, maybe it's build as a static lib on Windows?

The reason why I asked is that I actually have begun to work on fastrtps as well, it has even more issues like this one, but again, fastrtsp seems to be ready for windows shared dll because it already has all the code there, just not tested, maybe nobody wants to build it on Windows as shared dll?

**Comment by felixf4xu** [comment]
And I see the github actions for this fastcdr project, Windows is also listed there. I'm curious if it's configured to compile as static lib?

I'm not familiar with github actions, is it possible to set some actions to build on windows as shared dll?

**Comment by MiguelCompany** [comment]
@felixf4xu It might be something related with the compiler.

As stated in the supported platforms [here](https://fast-dds.docs.eprosima.com/en/stable/notes/versions.html#build-system-dependencies), our tier 1 build system on Windows is Visual Studio 2019.

On that tier 1 platform we pass the CI with `BUILD_SHARED_LIBS=ON`, so the DLL system is always tested.

The error you mention in #209 is usually reported by MinGW. If you are building in MinGW then bear in mind that it is not an officially supported platform.

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastCDR%20Manual%20Linux/400/)
* Mac Build Status
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastCDR%20Manual%20Windows/381/)

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastCDR%20Manual%20Linux/400/)
* Mac Build Status
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastCDR%20Manual%20Windows/382/)

**Comment by MiguelCompany** [comment]
@mergifyio rebase

**Comment by MiguelCompany** [comment]
CI:
* [Windows](https://github.com/eProsima/Fast-CDR/actions/runs/11773794535)
* [Ubuntu](https://github.com/eProsima/Fast-CDR/actions/runs/11773796344)
* [Mac](https://github.com/eProsima/Fast-CDR/actions/runs/11773797644)

**Comment by MiguelCompany** [comment]
CI run successfully on #247

**Pull request merged**